### PR TITLE
Internal affairs upgrades [READY]

### PR DESCRIPTION
**Pull request by 4dplanner**
:cl:
add: Internal Affairs Agents now obtain the kill objectives of their targets when they die.
add: Internal Affairs Agents now have an integrated nanotrasen pinpointer that tracks their target at distances further than ten squares.
add: Internal Affairs Agents will now lose any restrictions on collateral damage and gain a "Die a glorious death" objective upon becoming the last man standing, and revert if any of their targets are cloned.
/:cl:

This deals with two of the main problems with internal affairs, namely
1) nothing to do after objectives, and 
2) people hiding in lockers

If a target is cloned, any agents who have them as objectives will receive a warning, and lose any collateral privileges. The pinpointer will always point to the first target alive in your objective list.

Possible upgrades include giving the agents new objectives once they are the last person standing, and giving better messages in a lot of cases.

### CODE REVIEW

**Comment by optimumtact** [comment]
changelog name

**Comment by 4dplanner** [comment]
Fixed, thanks

**Comment by BeeSting12** [comment]
What if the target their target has is themselves?

**Comment by 4dplanner** [comment]
I've updated the main post with more information. If they obtain themselves as a target they lose their restrictions on collateral damage.

**Comment by 4dplanner** [comment]
Actually need to add an extra check to make sure that that doesn't happen when some of their targets are alive

**Comment by BeeSting12** [comment]
nice, this sounds really fun actually. i'd be willing to give the gamemode another chance with this merged

**Comment by Supermichael777** [comment]
ojgective X: cause a commotion and return to base.

**Comment by 4dplanner** [comment]
Something similar to that might work - return to base sort of shoehorns them into hijacking, though. I was thinking something more along the lines of "Sabotage Nanotrasen's operations  how you see fit" that auto-completes.

**Comment by PKPenguin321** [comment]
Really good, actually.

File: code/game/gamemodes/objective.dm [hunk 1]
```diff
@@ -695,6 +698,10 @@ GLOBAL_LIST_EMPTY(possible_items_special)
 		explanation_text = "Destroy [target.name], the experimental AI."
 	else
 		explanation_text = "Free Objective"
+	
+/datum/objective/destroy/internal
+	var/stolen = 0 		//Have we already eliminated this target?
+	var/traitored = 0	//Will we stop being the traitor if this target is killed?
```
Comment (ChangelingRain) [inline code/game/gamemodes/objective.dm]: TRUE/FALSE defines for all of these, please

**Comment by KorPhaeron** [comment]
Cool

**Comment by 4dplanner** [comment]
It should be, but it's 3 AM in England so I'd have to sort that tomorrow

**Comment by KorPhaeron** [comment]
Fair enough

File: code/game/gamemodes/traitor/double_agents.dm [hunk 2]
```diff
@@ -18,33 +18,134 @@
...
+/proc/steal_targets(datum/mind/owner,datum/mind/victim)
+	if(!owner.current||owner.current.stat==DEAD) //Should already be guaranteed if this is only called from steal_targets_timer_func, but better to be safe code than sorry code 
+		return
+	var/traitored = FALSE
+	var/failed_traitored = FALSE
+	for(var/objective_ in victim.objectives)
+		if(istype(objective_, /datum/objective/assassinate/internal))
+			var/datum/objective/assassinate/internal/objective = objective_
+			if(objective.target==owner)
+				to_chat(owner.current,"<B><font size=3 color=red> Now that all the loyalist agents have been purged, your syndicate sleeper training activates - YOU ARE THE TRAITOR! You now have no limits on collateral damage.</font></B>")
```
Comment (KorPhaeron) [inline code/game/gamemodes/traitor/double_agents.dm]: This is entirely subjective LORE nonsense but I'd rather it be something along the lines of "All the other agents are dead, and you're the last loose end. Stage a Syndicate terrorist attack to cover up for todays events. You no longer have any limits on collateral damage." 

And give them the "Die a glorious death" thing so they can just go nuts and keep their greentext.
 
But like I said subjective based on me not liking secret brainwashing changing your character motivations midway through, so maybe others should weigh in.

**Comment by KorPhaeron** [comment]
Also thank you for rescuing my half baked experiment

**Comment by Repukan** [comment]
That's pretty good actually

File: code/game/gamemodes/traitor/double_agents.dm (hunk 2 above)
Comment (ChangelingRain) [inline code/game/gamemodes/traitor/double_agents.dm]: I agree, swapping your motivations like this is weird and kor's suggestion feels better.

**Comment by demonfiren** [comment]
if they end up having to kill their own guys and fuck up their company's station to get out don't they kinda stop being internal affairs and become double agents again?

**Comment by Robustin** [comment]
I like the pinpointer variation, really nice idea, but here's what you should change:

1) Increase the "minimum distance" to ~15-20 from 10. It should give you an area to start looking around, 10 is basically "ok he's in xenobio" whereas 20 is more like "ok he's in science somewhere". Ideally this distance (when combined with a slower tick) should have moderate randomization so that each update (works better with slower updates described below) isn't perfect. Otherwise even at a larger range people will have a very precise idea of location when their pointer fades. This isn't as much of a problem for moving targets but targets who aren't constantly dashing from room to room, seeing your pointer fade at precisely X tiles would give experienced players an exact location for their target.

2) I'd also consider making the pinpointer update slowly, like with a new "ping" every 10 seconds or so. Constant updates would make it just a little too easy to piece together exactly where your target is. Given that IAA's can't just murder someone who seems a little shady, the hunters can't have too much of an advantage, there needs to be some uncertainty for both sides. 

3) IAA shouldn't activate upon "last man standing", inevitably people will fuck off to space/Lavaland, or the last man standing will metagame and start dumping their targets naked into a walled-off section of maint. Have it activate at 25% of the original number but disable the mechanic once the shuttle is past the point of no return. 

4) Your changes help keep the round interesting for the antags, but it still like the only purpose for non-antags is to sit around drooling at a wall hoping you win the antag reroll lottery. If you just sit around doing your job IAA's have no reason to mess with you and vice-versa. I see no reason not to throw a few actual traitors into the chain. They will still track targets and be tracked, but without the rule limitations that make IAA so benign. With this variation you would only need 20-33% of all IAA rolls to be traitors, just enough that non-antags aren't so safe and that the metagame doesn't solidify around "Oh its IAA time to stop giving a shit about people running around with Syndicate gear". 

5) I'd still like to see a system where the "agents" (even including the traitor-agents) start with 10TC and get something like 2-4TC per kill. I've made similar suggestions for DA simply because it would mitigate murderboning, but for IAA it would primarily be to add more variety to gameplay. Giving them all 20TC at roundstart just feels a little bit too much like "ebows + thermals in maint, Fox only, Final Destination". Starting with 10TC would encourage more creativity in how to go about eliminating targets. I'd also suggest curating the uplink list for IAA mode so that you can just remove boring/problematic choices from the mode.

**Comment by 4dplanner** [comment]
I'll put in 1 & 2 once I've finished testing the ability based pinpointers. 
3 is possible and might be good, but I'd want someone else to comment on whether it should be added.
4 & 5 sound like good ideas, but are against atomicity of PRs unless I've misunderstood, so would be something to do afterwards (though they would both be easy to add under this system).

**Comment by Robustin** [comment]
Sure I dont expect you to do everything at once. 

I have an alert-based tracker in my blood cult PR you could look at. To my knowledge its the only tracker that isn't based on organs or items, its simply a permanent alert thrown to someone when they get a certain antag status. Currently it works off matrices but the original version in #23866 was essentially a pinpointer alert.

**Comment by 4dplanner** [comment]
I'll take a look, thanks

**Comment by 4dplanner** [comment]
(I am aware of the various errors, and will fix them when I get back)

File: code/game/gamemodes/traitor/double_agents.dm [hunk 3]
```diff
@@ -18,33 +21,181 @@
...
+	scan_for_target()
+	point_to_target()
+	var/datum/callback/C = new(src, .pinpointer_ping_func)
+	addtimer(C, 100)
+
+/datum/objective/assassinate/internal/proc/give_pinpointer()
+	if(owner && owner.current)
+		var/datum/action/agent_pinpointer/pinp = new
+		pinp.Grant(owner.current)
+		pinp.pinpointer_ping_func()
```
Comment (Iamgoofball) [inline code/game/gamemodes/traitor/double_agents.dm]: you misspelled pimp

**Comment by 4dplanner** [comment]
I've probably missed a bunch of things, but it seemed to work fine in testing

**Comment by ChangelingRain** [comment]
Could you rephrase the "Internal Affairs Agents will now "activate" on becoming the last man standing" changelog entry, and add periods at the end of each entry?

File: code/game/gamemodes/traitor/double_agents.dm [hunk 4]
```diff
@@ -18,29 +22,179 @@
...
+				continue
+			if(objective.check_completion())
+				if(objective.stolen)
+					continue
+				else
+					steal_targets(objective.target)
+					objective.stolen = TRUE
+			else
+				if(objective.stolen)
+					var/fail_msg = "<B><font size=3 color=red>Your sensors tell you that [objective.target.current.real_name], one of the targets you were meant to have killed, pulled one over on you, and is still alive - do the job properly this time! </font></B>"
```
Comment (ChangelingRain) [inline code/game/gamemodes/traitor/double_agents.dm]: instead of `<B><font size=3 color=red>`, use `<span class='userdanger'>`
do this for all places this appears.

File: code/game/gamemodes/traitor/double_agents.dm [hunk 5]
```diff
@@ -18,29 +22,179 @@
...
 		if(i + 1 > traitors.len)
 			i = 0
-		target_list[traitor] = traitors[i + 1]
+		target_list[traitor] = traitors[i+1]	
 	..()
 
+
+/datum/action/agent_pinpointer
+	name = "Internal Affairs Integrated Pinpointer"
+	desc = "Even stealthier than a normal implant"
```
Comment (ChangelingRain) [inline code/game/gamemodes/traitor/double_agents.dm]: period.

File: code/game/gamemodes/traitor/double_agents.dm [hunk 6]
```diff
@@ -18,29 +22,179 @@
...
 	for(var/datum/mind/traitor in traitors)
 		i++
 		if(i + 1 > traitors.len)
 			i = 0
-		target_list[traitor] = traitors[i + 1]
+		target_list[traitor] = traitors[i+1]	
 	..()
 
+
+/datum/action/agent_pinpointer
```
Comment (ChangelingRain) [inline code/game/gamemodes/traitor/double_agents.dm]: This could probably just be a /datum/status_effect instead, this is kinda a weird solution; status effects already have processing and can do everything this can.
Also, they don't appear on the action bar, which is a weird place for the pinpointer to show up in the first place.

File: code/game/gamemodes/traitor/double_agents.dm [hunk 7]
```diff
@@ -1,9 +1,13 @@
...
+#define PINPOINTER_PING_TIME 40
+
 /datum/game_mode/traitor/internal_affairs
 	name = "Internal Affairs"
 	config_tag = "internal_affairs"
 	employer = "Internal Affairs"
-	required_players = 25
-	required_enemies = 5
+	required_players = 1
+	required_enemies = 1
```
Comment (ChangelingRain) [inline code/game/gamemodes/traitor/double_agents.dm]: debug!

**Comment by 4dplanner** [comment]
From what I can see, status effects don't have inbuilt drawing at all - so would you rather they were drawn under alerts instead?

**Comment by 4dplanner** [comment]
Ignore me, for I am blind

**Comment by ChangelingRain** [comment]
Yeah, status effects use an alert type and you can just have that update.

**Comment by ChangelingRain** [comment]
It occurs to me that the status effect doesn't have a reference to the alert to update it, which is actually really not ideal at all and should probably be corrected, either by me or by you.

**Comment by 4dplanner** [comment]
I'll try and add that in as I go along, I suppose

File: code/game/gamemodes/traitor/double_agents.dm (hunk 6 above)
Comment (Iamgoofball) [inline code/game/gamemodes/traitor/double_agents.dm]: Is status effect code still hot garbage

File: code/game/gamemodes/traitor/double_agents.dm [hunk 8]
```diff
@@ -84,20 +84,16 @@
...
 	point_to_target()
-	var/datum/callback/C = new(src, .pinpointer_ping_func)
-	addtimer(C, PINPOINTER_PING_TIME)
 
 /proc/give_pinpointer(datum/mind/owner)
 	if(owner && owner.current)
-		var/datum/action/agent_pinpointer/pinp = new
-		pinp.Grant(owner.current)
-		pinp.pinpointer_ping_func()
+		new/datum/status_effect/agent_pinpointer(owner.current)
```
Comment (ChangelingRain) [inline code/game/gamemodes/traitor/double_agents.dm]: owner.current.apply_status_effect(/datum/status_effect/agent_pinpointer)

**Comment by ChangelingRain** [comment]
Let's try it!
